src/parse-mockdb.js

```diff
@@ -4,52 +4,175 @@ const Parse = require('parse-shim');
+  if (CONFIG.DEBUG) {
+    console.log(['[', ']'].join(prefix), JSON.stringify(object, null, 4));
+  }
 }
 
-const HANDLERS = {
-  GET: handleGetRequest,
-  POST: handlePostRequest,
-  PUT: handlePutRequest,
-  DELETE: handleDeleteRequest,
+function isOp(object) {
+  return object && typeof object === 'object' && '__op' in object;
 }
 
-var db = {};
-var hooks = {};
-var masks = {};
+function isPointer(object) {
+  return object && object.__type === 'Pointer';
+}
 
-var indirect = null;
-var outOfBandResults = null;
-var default_controller = null;
-var mocked = false;
+function isDate(object) {
+  return object && object.__type === 'Date';
+}
 
 /**
- * Mocks a Parse API server, by intercepting requests and storing/querying data locally
- * in an in-memory DB.
+ * Deserialize an encoded query parameter if necessary
  */
-function mockDB() {
-  if (!mocked) {
-    default_controller = Parse.CoreManager.getRESTController();
-    mocked = true;
-    Parse.CoreManager.setRESTController(MockRESTController);
+function deserializeQueryParam(param) {
+  if (!!param && (typeof param === 'object')) {
+    if (param.__type === 'Date') {
+      return new Date(param.iso);
+    }
   }
+  return param;
 }
 
 /**
- * Restores the original RESTController.
+ * Evaluates whether 2 objects are the same, independent of their representation
+ * (e.g. Pointer, Object)
  */
-function unMockDB() {
-  if (mocked) {
-    Parse.CoreManager.setRESTController(default_controller);
-    mocked = false;
+function objectsAreEqual(obj1, obj2) {
+  // scalar values (including null/undefined)
+  // eslint-disable-next-line eqeqeq
+  if (obj1 == obj2) {
+    return true;
+  }
+
+  // if any of those is null or undefined the other is not because
+  // of above --> abort
+  if (!obj1 || !obj2) {
+    return false;
+  }
+
+  // objects
+  if (_.isEqual(obj1, obj2)) {
+    return true;
+  }
+
+  // both pointers
+  if (obj1.objectId !== undefined && obj1.objectId === obj2.objectId) {
+    return true;
+  }
+
+  // search through array
+  if (Array.isArray(obj1)) {
+    return _.some(obj1, obj => objectsAreEqual(obj, obj2));
+  }
+
+  // both dates
+  if (isDate(obj1) && isDate(obj2)) {
+    return deserializeQueryParam(obj1) === deserializeQueryParam(obj2);
+  }
+
+  return false;
+}
+
+
+// Ensures `object` has an array at `key`. Creates array if `key` doesn't exist.
+// Will throw if value for `key` exists and is not Array.
+function ensureArray(object, key) {
+  if (!object[key]) {
+    object[key] = [];
+  }
+  if (!Array.isArray(object[key])) {
+    throw new Error("Can't perform array operation on non-array field");
   }
 }
 
+const MASKED_UPDATE_OPS = new Set(['AddRelation', 'RemoveRelation']);
+
+/**
+ * Operator functions assume binding to **object** on which update operator is to be applied.
+ *
+ * Params:
+ *    key   - value to be modified in bound object.
+ *    value - operator value, i.e. `{__op: "Increment", amount: 1}`
+ */
+const UPDATE_OPERATORS = {
```

Comment:
note that by switching to lambdas, bind(this) no worky.  I think its better this way, with an explicit first param of the object anyway.